filesystems/watcher.go

```diff
@@ -561,6 +561,55 @@ func parsecurrentOffsetBytes(name string, currentOffsetBytes []byte) int64 {
 	return currentOffsetInt
 }
 
+func (w *Watcher) concurrentUploadLogFile(ctx context.Context, logFileList []string) error {
+	// Iterate over each file and setup the work
+	var wg sync.WaitGroup
+	const concurrentUploaders = 8
+	var errs *multierror.Error
+
+	// Fill up a buffering channel, so we can drain slowly
+	uploadFileC := make(chan string, len(logFileList))
+	for _, j := range logFileList {
+		uploadFileC <- j
+	}
+	close(uploadFileC)
+
+	// How many workers do we need
+	uploadWorkers := concurrentUploaders
+	if len(logFileList) < uploadWorkers {
+		uploadWorkers = len(logFileList)
+	}
+	wg.Add(uploadWorkers)
+
+	// Define what the worker should do
+	uploadWork := func(ch <-chan string) {
+		defer wg.Done()
+		for logFile := range ch {
+			if ctx.Err() != nil {
+				errs = multierror.Append(errs, ctx.Err())
+				break
+			}
+			if CheckFileForStdio(logFile) {
+				continue
+			}
+			err := w.uploadLogfile(ctx, logFile, true)
+			errs = multierror.Append(errs, err)
+		}
+	}
+	// Start the workers
+	for i := 0; i < uploadWorkers; i++ {
+		go uploadWork(uploadFileC)
+	}
+
+	wg.Wait()
```

Comment:
So, a couple things:
1. One of the things you can use here is an errgroup: https://pkg.go.dev/golang.org/x/sync/errgroup - It will encapsulate the logic of tying together context cancellation and returning errors. There is also go-multierror's "Group": https://pkg.go.dev/github.com/hashicorp/go-multierror#Group
2. You cannot perform the line: 
```go
				errs = multierror.Append(errs, ctx.Err())
```

Safely, because it'll result in a concurency violation.

Reply:
Thanks for surfacing this! Using errgroup now.